system-addon/lib/TelemetryFeed.jsm

```diff
@@ -274,6 +278,32 @@ this.TelemetryFeed = class TelemetryFeed {
   }
 
   /**
+   * handlePagePrerendered - Set the session as prerendered
+   *
+   * @param  {string} portID the portID of the target session
+   */
+  handlePagePrerendered(portID) {
+    const session = this.sessions.get(portID);
+
+    if (!session) {
+      // It's possible the tab was never visible – in which case, there was no user session.
+      return;
+    }
+
+    session.perf.is_prerendered = true;
+  }
+
+  /**
+   * handleNewTabInit - Handle NEW_TAB_INIT
+   *
+   * @param  {obj} action the Action object
+   */
+  handleNewTabInit(action) {
+    const session = this.addSession(au.getPortIdOfSender(action), action.data.url);
+    session.perf.is_preloaded = action.data.browser.getAttribute("isPreloadBrowser") === "true";
```

Comment:
@krizsa This is getting called, indirectly, via a RemotePage:Init message handler in the main process.  Since I know this is removed by the redirect code from [bug 1376895](https://bugzilla.mozilla.org/show_bug.cgi?id=1376895#c5), I'm concerned that this could race, possibly, but not necessarily just, in the case where someone opens multiple tabs in a row.  Can you offer any insight into that whether that's true/possible?  If so, how would you feel about somehow persisting the isPreloadBrowser attribute on all tabs that were preloaded and haven't yet navigated away from about:newtab?  This would be super helpful for our telemetry...

Reply:
@dmose `isPreloadBrowser` attribute does stay until a different URI is loaded. Here's after opening some tabs very quickly:
```js
gBrowser.browsers.forEach(b => console.log(b.currentURI.spec, b.getAttribute("isPreloadBrowser"), b));

about:newtab true <browser type="content" message="true" messagemanagergroup="browsers" contextmenu="contentAreaContextMenu" tooltip="aHTMLTooltip" remoteType="web" remote="true" isPreloadBrowser="true" selectmenulist="ContentSelectDropdown" datetimepicker="DateTimePickerPanel" autoscrollpopup="autoscroller" nodefaultsrc="true" clickthrough="never" autocompletepopup="PopupAutoComplete"> 
about:newtab  <browser type="content" message="true" messagemanagergroup="browsers" contextmenu="contentAreaContextMenu" tooltip="aHTMLTooltip" remoteType="web" remote="true" autocompletepopup="PopupAutoComplete" selectmenulist="ContentSelectDropdown" datetimepicker="DateTimePickerPanel" autoscrollpopup="autoscroller" nodefaultsrc="true" clickthrough="never"> 
about:newtab true <browser type="content" message="true" messagemanagergroup="browsers" contextmenu="contentAreaContextMenu" tooltip="aHTMLTooltip" remoteType="web" remote="true" isPreloadBrowser="true" selectmenulist="ContentSelectDropdown" datetimepicker="DateTimePickerPanel" autoscrollpopup="autoscroller" nodefaultsrc="true" clickthrough="never" autocompletepopup="PopupAutoComplete"> 
about:newtab true <browser type="content" message="true" messagemanagergroup="browsers" contextmenu="contentAreaContextMenu" tooltip="aHTMLTooltip" remoteType="web" remote="true" isPreloadBrowser="true" selectmenulist="ContentSelectDropdown" datetimepicker="DateTimePickerPanel" autoscrollpopup="autoscroller" nodefaultsrc="true" clickthrough="never" autocompletepopup="PopupAutoComplete"> 
about:newtab true <browser type="content" message="true" messagemanagergroup="browsers" contextmenu="contentAreaContextMenu" tooltip="aHTMLTooltip" remoteType="web" remote="true" isPreloadBrowser="true" selectmenulist="ContentSelectDropdown" datetimepicker="DateTimePickerPanel" autoscrollpopup="autoscroller" nodefaultsrc="true" clickthrough="never" autocompletepopup="PopupAutoComplete"> 
about:newtab  <browser type="content" message="true" messagemanagergroup="browsers" contextmenu="contentAreaContextMenu" tooltip="aHTMLTooltip" remoteType="web" remote="true" autocompletepopup="PopupAutoComplete" selectmenulist="ContentSelectDropdown" datetimepicker="DateTimePickerPanel" autoscrollpopup="autoscroller" nodefaultsrc="true" clickthrough="never" primary="true">
```

Reply:
@Mardak: interesting!  When I was poking around with DOMi, I wasn't able to actually see that attribute, so perhaps I was inadvertantly doing it wrong.  That said, the code I'm specifically concerned about is http://searchfox.org/mozilla-central/source/browser/base/content/browser.js#1170

Reply:
@Mardak Which, it appears, might be the two cases in your example that don't have the "true" in them.

Reply:
The ones without "true" are probably because it wasn't preloaded. I opened the tabs quickly.

Reply:
After a bunch of discussion wit @krizsa it sounds like that's right, and the code should be usable as-is. The redirect code I was concerned about just removes the attribute and causes it to be re-evaluated (and potentially re-added if necessary) in the parent process, because the child process doesn't have the information to know whether it's a preloaded browser or not.

@krizsa has said he would like to have one more look...

Reply:
Yes the code should be usable as-is. There are several ways to navigate away from about:newtab with different code paths, and to catch all the cases I had to add browser.removeAttribute("isPreloadBrowser"); at two places: http://searchfox.org/mozilla-central/search?q=removeAttribute%28%22isPreloadBrowser%22%29%3B&path=
The above mentioned RedirectLoad is handling the case where the navigation was requested from the child process (click on a link on about:newtab / search on about:newtab). Once the attribute is removed it won't ever get readded (since from that point on we won't care any more if it was a preloaded browser or not).

Reply:
Ok, now I understand.  Thanks @krizsa and @Mardak for help clarifying.